spring-integration-http/src/main/java/org/springframework/integration/http/inbound/HttpRequestHandlingEndpointSupport.java

```diff
@@ -505,7 +503,23 @@ protected void onInit() throws Exception {
 
 			Message<?> reply = null;
 			if (this.expectReply) {
-				reply = this.sendAndReceiveMessage(message);
+				try {
+					reply = this.sendAndReceiveMessage(message);
+				}
+				catch (MessageTimeoutException e) {
+					if (this.statusCodeExpression != null) {
+						reply = getMessageBuilderFactory().withPayload(e.getMessage())
+									.setHeader(org.springframework.integration.http.HttpHeaders.STATUS_CODE,
+											evaluateHttpStatus())
```

Comment:
You mean return a different status code for a timeout, based on the input message? I suppose we could but do you really think it's really necessary? I think it's easier for the user to understand if the behavior is consistent between the inbound channel adapter and gateway with timeout.